instrumentation/rxjava-2.0/library/src/test/groovy/RxJava2Test.groovy

```diff
@@ -0,0 +1,301 @@
+      maybe.blockingGet()
+    }
+
+    then:
+    assertTraces(1) {
+      trace(0, 3) {
+        sortSpansByStart()
+        basicSpan(it, 0, "trace-parent")
+        basicSpan(it, 1, "addOne", span(0))
+        basicSpan(it, 2, "addTwo", span(0))
+      }
+    }
+  }
+
+  def "Publisher chain spans have the correct parents from subscription time '#name'"() {
+    when:
+    assemblePublisherUnderTrace {
+      // The "add one" operations in the publisher created here should be children of the publisher-parent
+      def publisher = publisherSupplier()
+
+      runUnderTrace("intermediate") {
+        if (publisher instanceof Maybe) {
+          return ((Maybe) publisher).map(addTwo)
+        } else if (publisher instanceof Flowable) {
+          return ((Flowable) publisher).map(addTwo)
+        }
+        throw new IllegalStateException("Unknown publisher type")
+      }
+    }
+
+    then:
+    assertTraces(1) {
+      trace(0, 3 + 2 * workItems) {
+        sortSpansByStart()
+        basicSpan(it, 0, "trace-parent")
+
+        basicSpan(it, 1, "publisher-parent", span(0))
+        basicSpan(it, 2, "intermediate", span(1))
+
+        for (int i = 3; i < 3 + 2 * workItems; i = i + 2) {
+          basicSpan(it, i, "addOne", span(1))
+          basicSpan(it, i + 1, "addTwo", span(1))
+        }
+      }
+    }
+
+    where:
+    name             | workItems | publisherSupplier
+    "basic maybe"    | 1         | { -> Maybe.just(1).map(addOne) }
+    "basic flowable" | 2         | { -> Flowable.fromIterable([1, 2]).map(addOne) }
+  }
+
+  def "Flowables produce the right number of results '#scheduler'"() {
+    when:
+    List<String> values = Flowable.fromIterable(Arrays.asList(1, 2, 3, 4))
+      .parallel()
+      .runOn(scheduler)
+      .flatMap({ num -> Maybe.just(num.toString() + " on " + Thread.currentThread().getName()).toFlowable() })
+      .sequential()
+      .toList()
+      .blockingGet()
+
+    then:
+    values.size() == 4
+
+    where:
+    scheduler << [Schedulers.newThread(), Schedulers.computation(), Schedulers.single(), Schedulers.trampoline()]
+  }
+
+  def cancelUnderTrace(def publisherSupplier) {
+    runUnderTrace("trace-parent") {
+      runUnderTraceWithoutExceptionCatch("publisher-parent") {
+        def publisher = publisherSupplier()
+        if (publisher instanceof Maybe) {
+          publisher = publisher.toFlowable()
+        }
+
+        publisher.subscribe(new Subscriber<Integer>() {
+          void onSubscribe(Subscription subscription) {
+            subscription.cancel()
+          }
+
+          void onNext(Integer t) {
+          }
+
+          void onError(Throwable error) {
+          }
+
+          void onComplete() {
+          }
+        })
+      }
+    }
+  }
+
+  def assemblePublisherUnderTrace(def publisherSupplier) {
+    runUnderTrace("trace-parent") {
+      // The "add two" operations below should be children of this span
+      runUnderTraceWithoutExceptionCatch("publisher-parent") {
```

Comment:
Why is this additional span needed? There's no error handling in rxjava instrumentation at all, not a single place tries to get the current span - so I don't really understand what this part is supposed to verify.

Reply:
When I was porting DD implementation I've noticed that this additional span helps in writing validations in `Publisher '#name' test`. I'll try to refactor tests so that the additional span will not be required.

Reply:
Additional span removed 👍